src/irs/core/storage_inspector.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Functionality to periodically deal with stale files in configured object storage."""
+
+import logging
+from datetime import timedelta
+
+from ghga_service_commons.utils.multinode_storage import S3ObjectStorages
+from ghga_service_commons.utils.utc_dates import now_as_utc
+from pydantic import Field
+from pydantic_settings import BaseSettings
+
+from irs.ports.inbound.storage_inspector import StorageInspectorPort
+from irs.ports.outbound.dao import StagingObjectDaoPort
+
+log = logging.getLogger(__name__)
+
+
+class StorageInspectorConfig(BaseSettings):
+    """Config specific to storage inspector"""
+
+    object_stale_after_minutes: int = Field(
+        default=...,
+        description="Amount of time in minutes after which an object in the staging bucket"
+        + " is considered stale. If an object continues existing after this point in time,"
+        + " this is an indication, that something might have gone wrong downstream.",
+    )
+
+
+class StagingInspector(StorageInspectorPort):
+    """Checks inbox storage buckets for stale files."""
+
+    def __init__(
+        self,
+        config: StorageInspectorConfig,
+        staging_object_dao: StagingObjectDaoPort,
+        object_storages: S3ObjectStorages,
+    ):
+        """Initialize with DB DAOs and storage handles."""
+        self._config = config
+        self._staging_object_dao = staging_object_dao
+        self._object_storages = object_storages
+
+    async def check_buckets(self):
+        """Check objects in all buckets configured for the service."""
+        log.debug("Checking for stale objects.")
+
+        async for staging_object in self._staging_object_dao.find_all(mapping={}):
```

Comment:
If for some reason there is a file in the storage but not in the staging object dao, it would be missed. Thus I would propose to iterate over a list of objects obtained from the object storage instead.

Reply:
My first implementation did it that way, but there's some overhead involved, i.e. stale objects are fetched from the DB for each storage node configured, but need to be ignored for non matching storage_aliases as mapping only works on a single field.

Probably won't matter much in practice, though, so changed it back to as requested.

Reply:
You should be able to use the mapping in the find method of the DAO to match multiple fields. Isn't this working?